### PR TITLE
MH-12091: Implement per-tenant digest user for capture agents

### PR DESCRIPTION
**Pull request by pmiddend**
This PR adds a new in-memory per-tenant user and a new role `ROLE_CAPTURE_AGENT` for capture agents. It builds upon work by @smarquard and @lkiesow, see the [bitbucket PR](https://bitbucket.org/opencast-community/opencast/pull-requests/1542).

The new user is able to digest-auth in order to ingest recordings and thus has to be an in-memory user with clear-text password. Restricting the user's rights turned out to be rather tricky because of a "somewhat circular" dependency between `AssetManagerWithMessaging` and `AssetManagerWithSecurity`.

I tried to preserve the older commits in order to see who introduced which feature.

This work is sponsored by SWITCH.

### CODE REVIEW

File: modules/common/src/main/java/org/opencastproject/security/api/SecurityConstants.java [hunk 1]
```diff
@@ -56,4 +59,7 @@
   /** The roles associated with the Opencast system account */
   String[] GLOBAL_SYSTEM_ROLES = new String[] { GLOBAL_ADMIN_ROLE, GLOBAL_SUDO_ROLE };
 
+  /** The roles associated with the Opencast capture agent account */
+  String[] GLOBAL_CA_ROLES = new String[] { GLOBAL_CAPTURE_AGENT_ROLE };
```
Comment (staubesv) [inline modules/common/src/main/java/org/opencastproject/security/api/SecurityConstants.java]: GLOBAL_CAPTURE_AGENT_ROLES

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 2]
```diff
@@ -65,12 +66,18 @@
 
   /** The digest user */
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: user_s_

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 3]
```diff
@@ -65,12 +66,18 @@
...
   public static final String DIGEST_USER_NAME = "System User";
+  public static final String CAPTURE_AGENT_USER_NAME = "Capture Agent";
 
   /** Configuration key for the digest user */
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: plural

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 4]
```diff
@@ -65,12 +66,18 @@
...
   public static final String DIGEST_USER_KEY = "org.opencastproject.security.digest.user";
+  public static final String CAPTURE_AGENT_USER_KEY = "org.opencastproject.security.captureagent.user";
 
   /** Configuration key for the digest password */
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: plural

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 5]
```diff
@@ -65,12 +66,18 @@
...
   public static final String DIGEST_PASSWORD_KEY = "org.opencastproject.security.digest.pass";
+  public static final String CAPTURE_AGENT_PASSWORD_KEY = "org.opencastproject.security.captureagent.pass";
+
+  /** Configuration key for optional additional roles for the CA user */
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: capture agent user

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 6]
```diff
@@ -114,19 +121,50 @@ private void createSystemUsers() {
...
       }
 
-      // Create the digest user
+      // Role set for the capture agent user
+      Set<JaxbRole> caRoleList = new HashSet<JaxbRole>();
+      for (String roleName : SecurityConstants.GLOBAL_CA_ROLES) {
+        caRoleList.add(new JaxbRole(roleName, jaxbOrganization));
+      }
+
+      // Add the organization anonymous role to the CA user
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: capture agent user

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 7]
```diff
@@ -114,19 +121,50 @@ private void createSystemUsers() {
...
+      caRoleList.add(new JaxbRole(organization.getAnonymousRole(), jaxbOrganization));
+
+      String caExtraRoles = organization.getProperties().get(CAPTURE_AGENT_EXTRA_ROLES_KEY);
+      String caUsername = organization.getProperties().get(CAPTURE_AGENT_USER_KEY);
+      // Add any extra custom roles to the CA user
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: capture agent user

File: etc/custom.properties [hunk 8]
```diff
@@ -42,6 +42,13 @@ org.opencastproject.admin.email=admin@localhost
 org.opencastproject.security.digest.user=opencast_system_account
 org.opencastproject.security.digest.pass=CHANGE_ME
 
+# The username and password for capture agents.
+#org.opencastproject.security.ca.user=opencast_ca
+#org.opencastproject.security.ca.pass=CHANGE_ME
+
+# Optional extra roles for the capture agent user. These roles may be used to manage which series are visible to the CA.
```
Comment (staubesv) [inline etc/custom.properties]: [...] visible to the capture agent

File: etc/custom.properties [hunk 9]
```diff
@@ -42,6 +42,13 @@ org.opencastproject.admin.email=admin@localhost
 org.opencastproject.security.digest.user=opencast_system_account
 org.opencastproject.security.digest.pass=CHANGE_ME
 
+# The username and password for capture agents.
+#org.opencastproject.security.ca.user=opencast_ca
```
Comment (staubesv) [inline etc/custom.properties]: captureagent.user

File: modules/common/src/main/java/org/opencastproject/security/api/SecurityConstants.java [hunk 10]
```diff
@@ -47,6 +47,9 @@
   /** Name of the Opencast admin role */
   String GLOBAL_ADMIN_ROLE = "ROLE_ADMIN";
 
+  /** Name of the Opencast Capture Agent role */
```
Comment (staubesv) [inline modules/common/src/main/java/org/opencastproject/security/api/SecurityConstants.java]: capture agent

File: modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java [hunk 11]
```diff
@@ -114,19 +121,50 @@ private void createSystemUsers() {
...
+      // Create the system user
       if (digestUsername != null && digestUserPass != null) {
-        logger.info("Creating the system digest user");
+        logger.info("Creating the system digest user '{}'", digestUsername);
         User digestUser = new JaxbUser(digestUsername, digestUserPass, DIGEST_USER_NAME, null, getName(), true,
                 jaxbOrganization, roleList);
         inMemoryUsers.add(digestUser);
       }
+
+      String caUserPass = organization.getProperties().get(CAPTURE_AGENT_PASSWORD_KEY);
```
Comment (staubesv) [inline modules/userdirectory/src/main/java/org/opencastproject/userdirectory/InMemoryUserAndRoleProvider.java]: Can you please:

- Move the logic to create the capture agent user (incl. extra roles) to one single place
- First check caUsername != null && caUserPass != null and then do the logic to add extra roles

File: etc/security/mh_default_org.xml [hunk 12]
```diff
@@ -249,7 +249,8 @@
     <sec:intercept-url pattern="/export/**" method="GET" access="ROLE_ANONYMOUS" />
 
     <!-- Enable anonymous access to the annotation and the series endpoints -->
-    <sec:intercept-url pattern="/series/**" method="GET" access="ROLE_ANONYMOUS" />
+    <sec:intercept-url pattern="/series/**" method="GET" access="ROLE_ANONYMOUS, ROLE_CAPTURE_AGENT" />
+    <sec:intercept-url pattern="/assets/**" method="GET" access="ROLE_ADMIN, ROLE_CAPTURE_AGENT" />
```
Comment (lkiesow) [inline etc/security/mh_default_org.xml]: Do we explicitly need to add `ROLE_ADMIN`? Shouldn't that always work?
Why does a capture agent need read access to assets?

File: etc/security/mh_default_org.xml [hunk 13]
```diff
@@ -264,11 +265,19 @@
     <sec:intercept-url pattern="/feeds/**" method="GET" access="ROLE_ANONYMOUS" />
 
     <!-- Secure the system management URLs for admins only -->
-    <sec:intercept-url pattern="/services/*" access="ROLE_ADMIN" />
+    <sec:intercept-url pattern="/services/*" method="PUT" access="ROLE_ADMIN" />
```
Comment (gregorydlogan) [inline etc/security/mh_default_org.xml]: The PUT, POST, and DELETE intercepts here are redundant, and caught by the catch-all on (currently) line 299.

File: modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithMessaging.java [hunk 14]
```diff
@@ -229,9 +234,9 @@ public void notifyDeleteEpisode(String mpId) {
    * <p>
    * Do not call outside of a security context.
    */
-  private TakeSnapshot mkTakeSnapshotMessage(Snapshot snapshot) {
-    final AccessControlList acl = authSvc.getActiveAcl(snapshot.getMediaPackage()).getA();
-    return AssetManagerItem.add(workspace, snapshot.getMediaPackage(), acl, getVersionLong(snapshot),
+  private TakeSnapshot mkTakeSnapshotMessage(Snapshot snapshot, MediaPackage mp) {
+    final MediaPackage chosenMp = mp != null ? mp : snapshot.getMediaPackage();
```
Comment (gregorydlogan) [inline modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithMessaging.java]: This type of conditional is frowned upon in the codebase, Entwine's initial code notwithstanding.

File: modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurity.java [hunk 15]
```diff
@@ -97,15 +92,11 @@ private boolean isAuthorizedByAcl(Version version, String mpId, String action) {
...
-    if (isAuthorizedByAcl(mp, WRITE_ACTION)) {
-      final Snapshot snapshot = super.takeSnapshot(owner, mp);
-      final AccessControlList acl = authSvc.getActiveAcl(mp).getA();
-      storeAclAsProperties(snapshot, acl);
-      return snapshot;
-    } else {
-      return chuck(new UnauthorizedException("Not allowed to take snapshot of media package " + mp.getIdentifier().toString()));
-    }
+  @Override public Snapshot takeSnapshot(String owner, MediaPackage mp) {
+    final AccessControlList acl = authSvc.getActiveAcl(mp).getA();
```
Comment (gregorydlogan) [inline modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurity.java]: Previously the permissions were checked here, now they're not.  Considering this is the layer where security is supposed to be enforced, why this change?

File: modules/asset-manager-impl/src/test/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurityTest.java [hunk 16]
```diff
@@ -168,13 +153,13 @@ private Object parametersForTestTakeSnapshot() {
     final Organization org = TestOrganization.mkDefault();
     return $a($a(acl(ace(ROLE_TEACHER, WRITE_ACTION)),
                  TestUser.mk(org, ROLE_USER),
-                 false),
+                 true),
```
Comment (gregorydlogan) [inline modules/asset-manager-impl/src/test/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurityTest.java]: This should still be false?  Your ACL is ROLE_TEACHER, and you're trying to take the snapshot as ROLE_USER...

File: modules/asset-manager-impl/src/test/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurityTest.java [hunk 17]
```diff
@@ -168,13 +153,13 @@ private Object parametersForTestTakeSnapshot() {
...
               $a(acl(ace(ROLE_TEACHER, WRITE_ACTION)),
                  TestUser.mk(org, ROLE_USER, ROLE_TEACHER),
                  true),
               $a(acl(ace(ROLE_TEACHER, WRITE_ACTION)),
                  TestUser.mk(org),
-                 false),
+                 true),
```
Comment (gregorydlogan) [inline modules/asset-manager-impl/src/test/java/org/opencastproject/assetmanager/impl/AssetManagerWithSecurityTest.java]: This should still be false?  Your ACL is ROLE_TEACHER, and you're trying to take the snapshot as ROLE_ANONYMOUS (I think?)

**Review by gregorydlogan — CHANGES_REQUESTED** [review]

File: etc/security/mh_default_org.xml (hunk 12 above)
Comment (pmiddend) [reply etc/security/mh_default_org.xml]: Good catch. The `assets` rule was just for debugging purposes. I'll remove that.

File: etc/security/mh_default_org.xml (hunk 13 above)
Comment (pmiddend) [reply etc/security/mh_default_org.xml]: Strangely enough, when I remove the three lines (PUT, POST, DELETE), ingesting doesn't work anymore (forbidden). I'll need to investigate this further.

File: modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithMessaging.java (hunk 14 above)
Comment (pmiddend) [reply modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerWithMessaging.java]: Didn't know that, I'll rewrite it, then.

File: etc/security/mh_default_org.xml (hunk 13 above)
Comment (staubesv) [reply etc/security/mh_default_org.xml]: Hi @pmiddend, can you please replace those lines by

```
    <sec:intercept-url pattern="/services/available.*" method="GET" access="ROLE_ADMIN, ROLE_CAPTURE_AGENT" />
    <sec:intercept-url pattern="/services/**" access="ROLE_ADMIN"/>
```

@gregorydlogan Line 299 does also provide access to ROLE_COURSE_ADMIN so adding /services/* for ROLE_ADMIN here is not redundant

**Comment by staubesv** [comment]
Hi @gregorydlogan @pmiddend has updated the PR. Are there any open comments left?

File: modules/search-service-impl/src/main/java/org/opencastproject/search/impl/SearchServiceImpl.java [hunk 18]
```diff
@@ -357,14 +356,8 @@ public Job add(MediaPackage mediaPackage) throws SearchException, MediaPackageEx
    * @throws UnauthorizedException
    *           if the user does not have the rights to add the mediapackage
    */
-  public void addSynchronously(MediaPackage mediaPackage) throws SearchException, MediaPackageException,
+  private void addSynchronously(MediaPackage mediaPackage) throws SearchException, MediaPackageException,
           IllegalArgumentException, UnauthorizedException {
-    User currentUser = securityService.getUser();
```
Comment (gregorydlogan) [inline modules/search-service-impl/src/main/java/org/opencastproject/search/impl/SearchServiceImpl.java]: What's the reasoning behind this removal?  I *think* it's still safe since the relevant endpoint is only accessible by ROLE_ADMIN and ROLE_COURSE_ADMIN, but it seems like this check wouldn't hurt?  Why does the CA need to be able to talk to the search service at all?

**Comment by gregorydlogan** [comment]
One final question and we'll be done I think.

**Comment by pmiddend** [comment]
@gregorydlogan You're right about the `SearchService`. I somewhat blindly took this snipped from Lars' WIP branch for the problem. I removed it, ingested something with the CA role and it still worked. :)

**Review by gregorydlogan — APPROVED** [review]
